manager/controlapi/network.go

```diff
@@ -90,20 +94,45 @@ func validateNetworkSpec(spec *api.NetworkSpec, pg plugingetter.PluginGetter) er
 	return nil
 }
 
+func isIngressNetwork(n *api.Network) bool {
+	if n.Spec.Ingress {
+		return true
+	}
+	// Check if legacy defined ingress network
+	_, ok := n.Spec.Annotations.Labels["com.docker.swarm.internal"]
+	return ok && n.Spec.Annotations.Name == "ingress"
+}
+
+func (s *Server) validateNetworkRequest(spec *api.NetworkSpec, pg plugingetter.PluginGetter) error {
+	if err := validateNetworkSpec(spec, pg); err != nil {
+		return err
+	}
+
+	// We can only accept one Ingress network
+	if spec.Ingress {
+		if spec.DriverConfig.Name != "overlay" {
+			return grpc.Errorf(codes.Unimplemented, "only overlay driver is currently supported for ingress network")
+		}
+		n, err := allocator.GetIngressNetwork(s.store)
+		if err == nil {
+			return grpc.Errorf(codes.PermissionDenied, "ingress network (%s) is already present", n.ID)
+		}
+		if err != nil && grpc.Code(err) != codes.NotFound {
+			return err
+		}
+	}
+
+	return nil
+}
+
 // CreateNetwork creates and returns a Network based on the provided NetworkSpec.
 // - Returns `InvalidArgument` if the NetworkSpec is malformed.
 // - Returns an error if the creation fails.
 func (s *Server) CreateNetwork(ctx context.Context, request *api.CreateNetworkRequest) (*api.CreateNetworkResponse, error) {
-	// if you change this function, you have to change createInternalNetwork in
-	// the tests to match it (except the part where we check the label).
-	if err := validateNetworkSpec(request.Spec, s.pg); err != nil {
+	if err := s.validateNetworkRequest(request.Spec, s.pg); err != nil {
```

Comment:
This should be called inside the `s.store.Update` call below, to make sure multiple `CreateNetwork` invocations can't race to create multiple ingress networks.

You might consider changing `GetIngressNetwork` to take a `store.ReadTx` instead of a store.